### DISCUSSION TITLE
`get_model_definitions` does not correctly parse schema when `allow_population_by_field_name` is `True`

### DISCUSSION THREAD

**Discussion by jpribyl** [fastapi/fastapi#8751] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from pydantic import BaseModel, Field
from fastapi import FastAPI

class RequestBody(BaseModel):
    class Config:
        allow_population_by_field_name = True

    id: str = Field(..., alias="_id")


app = FastAPI()


@app.put("/")
def read_root(body: RequestBody):
    return {"Hello": "World"}
```

### Description

* Open the browser and call the endpoint `/` with body { "id": "test" }.
* It returns a JSON with `{"Hello": "World"}`.
* Open the browser and call the endpoint `/` with body { "_id": "test" }.
* It returns a JSON with `{"Hello": "World"}`.
* Open the browser and go to `/docs`
* It shows body with { "_id": str }
* I expect it to include possibility of { "id": str }

### Environment

* OS: [e.g. Linux / Windows / macOS]: archlinux
* FastAPI Version [e.g. 0.3.0]: fastapi==0.65.1
* Python version: 3.9.6

### Additional context

I was able to trace it down to this function:
https://github.com/tiangolo/fastapi/blob/22528373bba6a654323de416ad5c867cbadb81bb/fastapi/utils.py#L17-L30

Workaround is to override it like this (I want to prioritize "id" over "_id" but you could definitely merge them instead):
```python
def get_model_definitions(
    *,
    flat_models: Set[Union[Type[BaseModel], Type[Enum]]],
    model_name_map: Dict[Union[Type[BaseModel], Type[Enum]], str],
) -> Dict[str, Any]:
    definitions: Dict[str, Dict[str, Any]] = {}
    for model in flat_models:

        m_schema, m_definitions, m_nested_models = model_process_schema(
            model,
            model_name_map=model_name_map,
            ref_prefix=REF_PREFIX,
            by_alias=not model.Config.allow_population_by_field_name
            if issubclass(model, BaseModel)
            else True,
        )
        definitions.update(m_definitions)
        model_name = model_name_map[model]
        definitions[model_name] = m_schema
    return definitions
```

**Comment by jpribyl** [comment]
I think this is likely better fixed on the Pydantic side of things, but I'll leave this open for now as I can see an argument either way

**Comment by whiskeyriver** [comment, marked as answer]
@jpribyl 

I believe the way that FastAPI and Pydantic handle this is correct, and the example you give is actually a good illustration of this.

In your example you've given the `id` attribute a name that collides with a python builtin. You should be giving the attribute a correct name, like `id_`, and the alias should be "id". Then, with `allow_population_by_field_name = True` you can instantiate the class in your own code using the attribute name `id_`, but the OpenAPI will correctly show `id` as the field.

Consider the python snake-case vs. json camel-case disparity for your models.

```python
class PostUserModel(BaseModel):
    class Config:
        allow_population_by_field_name = True

    first_name: str = Field(..., alias="firstName")
    last_name: str = Field(..., alias="lastName")
    

@app.put("/users", response_model=PostUserModel)
def create_user(data: PostUserModel):
     logger.info("Creating user %s %s", data.first_name, data.last_name)
     return data

def test_create_user(client):
    user = PostUserModel(first_name="John", last_name="Doe")
    response = client.put("/users", json=user.dict(by_alias=True)
```

In your own code you access the attributes by their python name, and the schema is generated using the JSON name.
